app/src/main/kotlin/compiler/intermediate/generators/DefaultGeneratorDetailsGenerator.kt

```diff
@@ -0,0 +1,250 @@
+        cfgBuilder.mergeConditionally(
+            mainBody,
+            ControlFlowGraphBuilder().apply { addSingleTree(IFTNode.JumpToRegister(savedStateRegister)) }.build()
+        )
+
+        // epilogue
+
+        // set RDX to signalize generator end
+        cfgBuilder.addSingleTree(IFTNode.RegisterWrite(Register.RDX, IFTNode.Const(0)))
+
+        // restore callee-saved registers
+        val restoreCalleeSavedRegistersCFG = innerFDG.genCalleeSavedRegistersRestore()
+        cfgBuilder.mergeUnconditionally(restoreCalleeSavedRegistersCFG)
+
+        // restore display entry
+        cfgBuilder.mergeUnconditionally(innerFDG.genDisplayRestore())
+
+        // restore base pointer
+        cfgBuilder.addSingleTree(IFTNode.RegisterWrite(Register.RBP, IFTNode.StackPop()))
+
+        // return
+        cfgBuilder.addSingleTree(
+            IFTNode.Return(
+                setOf(Register.RAX, Register.RDX) + calleeSavedRegistersWithoutRSPAndRBP
+            )
+        )
+
+        // yield target
+        val yieldTarget = IFTNode.LabeledNode(".yield", IFTNode.RegisterWrite(Register.RDX, IFTNode.StackPop()))
+        cfgBuilder.addAllFrom(ControlFlowGraphBuilder().apply { addSingleTree(yieldTarget) }.build())
+        cfgBuilder.addLink(Pair(yieldTarget, CFGLinkType.UNCONDITIONAL), restoreCalleeSavedRegistersCFG.entryTreeRoot!!, false)
+
+        return cfgBuilder.build()
+    }
+
+    override fun genYield(value: IFTNode): ControlFlowGraph {
+        val cfgBuilder = ControlFlowGraphBuilder()
+
+        cfgBuilder.addSingleTree(IFTNode.RegisterWrite(Register.RAX, value))
+        cfgBuilder.addSingleTree(
+            IFTNode.Call(
+                IFTNode.MemoryLabel(".yield"),
+                listOf(Register.RAX),
+                callerSavedRegisters + calleeSavedRegistersWithoutRSPAndRBP
+            )
+        )
+
+        return cfgBuilder.build()
+    }
+
+    private fun getNestedForeachFramePointerAddressWithCustomBase(foreachLoop: Statement.ForeachLoop, base: Register): IFTNode? {
+        val index = nestedForeachLoops.indexOf(Ref(foreachLoop))
+        return if (index != -1)
+            IFTNode.Subtract(
+                IFTNode.RegisterRead(base),
+                IFTNode.Const(SummedConstant((index + 1) * memoryUnitSize.toLong(), innerFDG.requiredMemoryBelowRBP))
+            )
+        else null
+    }
+
+    override fun getNestedForeachFramePointerAddress(foreachLoop: Statement.ForeachLoop): IFTNode? =
+        getNestedForeachFramePointerAddressWithCustomBase(foreachLoop, Register.RBP)
+
+    override fun genFinalize(): ControlFlowGraph {
+        val cfgBuilder = ControlFlowGraphBuilder()
+
+        // generate the prologue
+        cfgBuilder.mergeUnconditionally(finalizeFDG.genPrologue())
+
+        // prepare custom base
+        val customBaseRegister = Register.R15 // R15 is callee-saved FIXME: consider using a virtual register here
```

Comment:
To improve this, we'd need to modify coloring to allow it picking arbitrary `calee-saved` register there, right? We can create an issue for that after merging.

Reply:
I think it could work without modification in coloring. This register will be in conflict with all caller-saved ones (because of CALL's), so the allocator will either give it callee-saved register (which is fine) or spill it (which should also be fine I guess). But I'm not sure about that and I can't test this class at the moment, so I did it a safer way.